File: backend/src/main/java/com/technicalescaperoom/backend/controller/admin/AdminTeamController.java
package com.technicalescaperoom.backend.controller.admin;

import com.technicalescaperoom.backend.dto.admin.*;
import com.technicalescaperoom.backend.service.TeamService;
import com.technicalescaperoom.backend.service.admin.TeamExcelImportService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

@RestController
@RequestMapping("/api/admin")
@RequiredArgsConstructor
public class AdminTeamController {

    private final TeamService teamService;
    private final TeamExcelImportService teamExcelImportService;

    @PostMapping("/events/{eventId}/teams")
    public ResponseEntity<TeamDetailResponse> createTeam(
            @PathVariable Long eventId,
            @Valid @RequestBody CreateTeamRequest request) {
        TeamDetailResponse response = teamService.createTeam(eventId, request);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/teams/{teamId}")
    public ResponseEntity<TeamDetailResponse> getTeam(@PathVariable Long teamId) {
        TeamDetailResponse response = teamService.getTeamById(teamId);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/events/{eventId}/teams")
    public ResponseEntity<List<TeamResponse>> listTeamsForEvent(@PathVariable Long eventId) {
        List<TeamResponse> response = teamService.getTeamsByEventId(eventId);
        return ResponseEntity.ok(response);
    }

    @PutMapping("/teams/{teamId}")
    public ResponseEntity<TeamDetailResponse> updateTeam(
            @PathVariable Long teamId,
            @Valid @RequestBody UpdateTeamRequest request) {
        TeamDetailResponse response = teamService.updateTeam(teamId, request);
        return ResponseEntity.ok(response);
    }

    @PatchMapping("/teams/{teamId}/status")
    public ResponseEntity<TeamDetailResponse> updateTeamStatus(
            @PathVariable Long teamId,
            @Valid @RequestBody UpdateTeamStatusRequest request) {
        TeamDetailResponse response = teamService.updateTeamStatus(teamId, request.getStatus());
        return ResponseEntity.ok(response);
    }

    private final com.technicalescaperoom.backend.service.admin.AdminDashboardService adminDashboardService;
    private final com.technicalescaperoom.backend.service.admin.AdminTeamResetService adminTeamResetService;
    private final com.technicalescaperoom.backend.service.PlayerSessionService playerSessionService;

    @GetMapping("/events/{eventId}/teams/progress")
    public ResponseEntity<List<AdminTeamProgressDto>> getTeamsProgress(
            @PathVariable Long eventId,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) Integer level,
            @RequestParam(required = false) String status) {
        List<AdminTeamProgressDto> response = adminDashboardService.getTeamsProgress(eventId, search, level, status);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/teams/{teamId}/reset")
    public ResponseEntity<Void> resetTeam(
            @PathVariable Long teamId,
            @org.springframework.security.core.annotation.AuthenticationPrincipal com.technicalescaperoom.backend.config.security.AdminPrincipal principal) {
        adminTeamResetService.resetTeamProgress(principal, teamId);
        return ResponseEntity.ok().build();
    }

    @PostMapping("/teams/{teamId}/pause")
    public ResponseEntity<TeamDetailResponse> pauseTeam(
            @PathVariable Long teamId,
            @org.springframework.security.core.annotation.AuthenticationPrincipal com.technicalescaperoom.backend.config.security.AdminPrincipal principal) {
        TeamDetailResponse response = teamService.updateTeamStatus(teamId, com.technicalescaperoom.backend.enums.TeamStatus.REGISTERED);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/teams/{teamId}/resume")
    public ResponseEntity<TeamDetailResponse> resumeTeam(
            @PathVariable Long teamId,
            @org.springframework.security.core.annotation.AuthenticationPrincipal com.technicalescaperoom.backend.config.security.AdminPrincipal principal) {
        TeamDetailResponse response = teamService.updateTeamStatus(teamId, com.technicalescaperoom.backend.enums.TeamStatus.REGISTERED);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/sessions/{sessionId}/revoke")
    public ResponseEntity<Void> revokeSession(
            @PathVariable Long sessionId,
            @org.springframework.security.core.annotation.AuthenticationPrincipal com.technicalescaperoom.backend.config.security.AdminPrincipal principal) {
        playerSessionService.revokeSession(principal, sessionId);
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/teams/{teamId}")
    public ResponseEntity<Void> deleteTeam(@PathVariable Long teamId) {
        teamService.deleteTeam(teamId);
        return ResponseEntity.noContent().build();
    }

    // ---- Excel Team Import Endpoints ----

    @PostMapping("/events/{eventId}/teams/import/preview")
    public ResponseEntity<TeamImportPreviewDto> previewExcelImport(
            @PathVariable Long eventId,
            @RequestParam("file") MultipartFile file) {
        TeamImportPreviewDto preview = teamExcelImportService.parseAndValidate(eventId, file);
        return ResponseEntity.ok(preview);
    }

    @PostMapping("/events/{eventId}/teams/import/confirm")
    public ResponseEntity<TeamImportResultDto> confirmExcelImport(
            @PathVariable Long eventId,
            @RequestParam("file") MultipartFile file,
            @org.springframework.security.core.annotation.AuthenticationPrincipal com.technicalescaperoom.backend.config.security.AdminPrincipal principal) {
        TeamImportResultDto result = teamExcelImportService.importTeams(eventId, file, principal);
        return ResponseEntity.status(HttpStatus.CREATED).body(result);
    }
}

File: backend/controllers/admin/adminTeamController.js
const teamService = require('../../services/teamService');
const teamExcelImportService = require('../../services/admin/teamExcelImportService');
const adminDashboardService = require('../../services/admin/adminDashboardService');
const adminTeamResetService = require('../../services/admin/adminTeamResetService');
const playerSessionService = require('../../services/playerSessionService');

class AdminTeamController {
  async createTeam(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const response = await teamService.createTeam(eventId, req.body);
      res.status(201).json(response);
    } catch (err) {
      next(err);
    }
  }

  async getTeam(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const response = await teamService.getTeamById(teamId);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async listTeamsForEvent(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const response = await teamService.getTeamsByEventId(eventId);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async updateTeam(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const response = await teamService.updateTeam(teamId, req.body);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async updateTeamStatus(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const response = await teamService.updateTeamStatus(teamId, req.body.status);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async getTeamsProgress(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const search = req.query.search;
      const level = req.query.level ? parseInt(req.query.level, 10) : null;
      const status = req.query.status;
      const response = await adminDashboardService.getTeamsProgress(eventId, search, level, status);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async resetTeam(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      await adminTeamResetService.resetTeamProgress(req.admin, teamId);
      res.json({ message: 'Team reset successfully' });
    } catch (err) {
      next(err);
    }
  }

  async pauseTeam(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const response = await teamService.updateTeamStatus(teamId, 'REGISTERED');
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async resumeTeam(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const response = await teamService.updateTeamStatus(teamId, 'REGISTERED');
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async revokeSession(req, res, next) {
    try {
      const sessionId = parseInt(req.params.sessionId, 10);
      await playerSessionService.revokeSession(req.admin, sessionId);
      res.json({ message: 'Session revoked' });
    } catch (err) {
      next(err);
    }
  }

  async getActiveSessions(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const response = await adminDashboardService.getActiveSessions(eventId);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async resetTeamCredentials(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      await playerSessionService.resetTeamCredentialsAndSessions(req.admin, teamId);
      res.json({ message: 'Team credentials and sessions reset' });
    } catch (err) {
      next(err);
    }
  }

  async revokeTeamSessions(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      await playerSessionService.revokeAllTeamSessions(req.admin, teamId);
      res.json({ message: 'Team sessions revoked' });
    } catch (err) {
      next(err);
    }
  }

  async resetAllSessions(req, res, next) {
    try {
      await playerSessionService.resetAllSessionsAndCredentials(req.admin);
      res.json({ message: 'All sessions reset' });
    } catch (err) {
      next(err);
    }
  }

  async deleteTeam(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      await teamService.deleteTeam(teamId);
      res.status(204).send();
    } catch (err) {
      next(err);
    }
  }

  async previewExcelImport(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      if (!req.file) {
        return res.status(400).json({ error: 'No file uploaded' });
      }
      const preview = await teamExcelImportService.parseAndValidate(eventId, req.file.buffer, req.file.originalname);
      res.json(preview);
    } catch (err) {
      next(err);
    }
  }

  async confirmExcelImport(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      if (!req.file) {
        return res.status(400).json({ error: 'No file uploaded' });
      }
      const result = await teamExcelImportService.importTeams(eventId, req.file.buffer, req.file.originalname, req.admin);
      res.status(201).json(result);
    } catch (err) {
      next(err);
    }
  }
}

module.exports = new AdminTeamController();
